### PR TITLE
executor: fix show grants failed after drop granted role

### LINKED ISSUE #29473: show grants failed after drop granted role
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```sql
-- session1
create user u1;
create role r1;
grant r1 to u1;
grant create user on *.* to 'u1';

-- session2, login in with u1
set role r1;
drop role r1;
show grants;
```

### 2. What did you expect to see? (Required)
```sql
MySQL u1@127.0.0.1:(none)> show grants;
+--------------------------------------+
| Grants for User                      |
+--------------------------------------+
| GRANT CREATE USER ON *.* TO 'u1'@'%' |
+--------------------------------------+


```
### 3. What did you see instead (Required)
```sql
MySQL u1@127.0.0.1:(none)> show grants;
(3530, '`r1`@`%` is not granted to u1@%')
```
### 4. What is your TiDB version? (Required)
master, release-5.3
```sql
MySQL u1@127.0.0.1:(none)> select tidb_version()\G
***************************[ 1. row ]***************************
tidb_version() | Release Version: v5.3.0-alpha-1301-g091cd29ca
Edition: Community
Git Commit Hash: 091cd29ca37b866bfd7a1eebedbfe8e5a1c4c1f1
Git Branch: release-5.3
UTC Build Time: 2021-11-04 09:18:04
GoVersion: go1.17.2
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false

```

### PR DESCRIPTION
**Pull request by unconsolable**
Signed-off-by: unconsolable <chenzhipeng2012@gmail.com>

### What problem does this PR solve?

Issue Number: close #29473

Problem Summary:

### What is changed and how it works?
- In executor of `DROP ROLE`, remove the role to be dropped from `activeRoles`
- Add a test case

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Fix show grants failed after drop granted role
```

### CODE REVIEW

**Comment by unconsolable** [comment]
/cc @morgo 
Please help on review when available, thanks~

File: executor/simple.go [hunk 1]
```diff
@@ -1296,13 +1296,30 @@ func (e *SimpleExec) executeDropUser(ctx context.Context, s *ast.DropUserStmt) e
 			break
 		}
 
+		// delete from activeRoles
+		if s.IsDropRole {
+			for i := 0; i < len(activeRoles); i++ {
+				if activeRoles[i].Username == user.Username && activeRoles[i].Hostname == user.Hostname {
+					activeRoles = append(activeRoles[:i], activeRoles[i+1:]...)
```
Comment (tiancaiamao) [inline executor/simple.go:1303]: Is the `activeRoles` information propagated to other TiDB instance?
If not, when you drop the role in one TiDB instance and query in another TiDB instance, the error remains there.

**Review by tiancaiamao — APPROVED** [review]
OK, at lease it solved part of the issue.

File: executor/simple.go (hunk 1 above)
Comment (unconsolable) [reply executor/simple.go:1303]: It is a problem, but I don't know how to solve it... I read through `(*SimpleExec).setRoleRegular` and can't find related logic.🤔

Comment (tiancaiamao) [reply executor/simple.go:1303]: ./executor/simple.go:   return domain.GetDomain(e.ctx).NotifyUpdatePrivilege()


I'm not sure whether this works. E.g, update the privilege cache to make every TiDB instance reload the active role info when drop user is called.

Comment (unconsolable) [reply executor/simple.go:1303]: I found it hard to do so. In `Domain` I can't find a way to get the related session, thus I don't know how to get the `activeRoles`. If we can get it, I think we can check and apply `activeRoles` in `(*Handle).Update()`.

**Review by morgo — APPROVED** [review]
LGTM

I think the integration test errors are unrelated.

**Comment by unconsolable** [comment]
@morgo Shall we keep the issue open to show the issue is not fully resolved?

**Comment by morgo** [comment]
> @morgo Shall we keep the issue open to show that it is not fully resolved?

I think we can say this issue is resolved but fork and create another issue. The privilege cache is also automatically re-generated every 5 minutes, so on a live cluster it will eventually resolve itself. Here is a similar case I fixed recently: https://github.com/pingcap/tidb/pull/27958 -- so there might be other cases as well.

**Pull request merged**